### ISSUE TITLE
Special characters in System generated passwords 

### ISSUE THREAD

**Issue by YannickBiet** [tracim/tracim#2798]
## Bug description & symptom

When creating a new account, the generated password is quite difficult to read/enter as it uses characters like `0 O ( { l I`.
System generated passwords should not use difficult characters to remain easy to write and to read.

Too many special characters or characters that look the same should be avoided. To do:
- [x] allow alphanumeric characters.
- [x] remove `0, l, 1, I `and `O` characters.
- [x] allow only `!#%@+=?` as special characters.
- [X] create/find a specific issue about changing the workflow : https://github.com/tracim/tracim/issues/3499

## Implemented solution
see https://github.com/tracim/tracim/pull/3508
- allow alphanumeric and `!#%@+=?` without "0Oo" and '1lI"' ambiguous characters as generated password characters
with explicit code to do this allowing understanding the logical process behind characters allowed/removed.

Original issue: 
---- 

> ## Bug description & symptom
> 
> System generated passwords should not use difficult characters to remain easily readable and acceptable by non IT-power-users
> 
> ## Diagnostic
> 
> @lebouquetin introduced Tracim as a powerful targeting both 
> - IT power user
> - also for non power users
> 
> When subscribing a new account on the system, the generated password is quite difficult to enter for non power user as it uses 0 O ( { l I (...) characters
> 
> ## Expected
> 
> Have strong password without too much exotic characters or "looking the same" characters
> 
> ## Context and version information
> 
> Mon.tracim.fr 2.6.2

-----
Changelog:
2020-09-02: updated issue with proposal about restricting characters.
2020-09-03: reformat the issue and remove the context that remains available in the quoted original issue
2020-09-04: updated issue with implemented solution

**Fixed in tracim 3.1.0**

**Comment by PhilippeAccorsi** [comment]
Another idea :
Not send password generated but a link to go on tracim page where user can define password manually. Form like we have for reset password.

In this cas, user can choose is own password and its better for user because its not necessary to go in profile account to change password after first connexion.

**Comment by YannickBiet** [comment]
depending on the complexity of your idea, limiting the range of caracters in round 1 could be a first good step.
I agree that changing the flow as you mentioned it is again better.

**Comment by inkhey** [comment]
Just checked code, allowed password character are currently these : 
https://github.com/tracim/tracim/blob/c237ed78cc54cbfd0231fd3e8743538bb90e0d41/backend/tracim_backend/lib/utils/utils.py#L157

Which mean theses characters:
`abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!#$%&()*+,-./:;=?@[]_{}`

The change of the workflow may be a best solution but require a bit more work for both backend and frontend side. This should not be a big deal as we have almost same mechanism for reset password.

**Comment by inkhey** [comment]
Todo:
- [ ] check good practices for password generation
- [ ] reduce char used in code
- [ ] create/check issue about better workflow.

**Comment by inkhey** [comment]
About good practice, i don't find clear answers about characters used for generated password but allowed characters in some case may help us found the most useful special characters for password generation:
What i've found : 
- IBM allow alphanumeric + !, @, #, $, %, ^, &, or * : https://www.ibm.com/support/pages/password-policy-and-passwords-special-characters
- https://superuser.com/questions/1197808/guidlines-for-special-characters-in-passwords#comment1741315_1197808 suggest that  !@#._  are commonly accepted.
- https://www.lastpass.com/fr/password-generator suggest to remove character 0, l, 1 and O which are ambiguous.
- https://lifehacker.com/how-to-create-secure-passwords-that-arent-impossible-to-1825048324 article show a view of lastpass with restricted list of special characters to just "!%@#" .

Nice publication about password on mobile:
https://tsapps.nist.gov/publication/get_pdf.cfm?pub_id=914910

I do suggest to:
- remove `0, l, 1, I `and `O` characters
- allow only `!#%@` as special characters, "`.;,`" are too small, brackets are confusing too, "`_`" is not very distingable too. $ is too country special i do think. "*^|~" are for power user. The others potential acceptable special characters may be `+=?`

**Comment by celiamgtt** [comment]
I agree with your suggestions.
For your question, allow only `!#%@+=?` looks fine for me :)

**Comment by PhilippeAccorsi** [comment]
Thanks for the investigation @inkhey 
I agree also with your suggestions.

**Comment by inkhey** [comment]
i updated the issue according to the new suggestion.

**Comment by inkhey** [comment]
to test this issue:
- try create new user with api and email notification activated but without setting password, email generated should contain only listed characters. User can be created both by admin user view or by inviting new user with email from workspace view.


To avoid use of real email, you can use mailhog:
```
email.notification.activated = True
email.notification.smtp.server = 127.0.0.1
email.notification.smtp.port = 1025
email.notification.smtp.user = test_user
email.notification.smtp.password = just_a_password
```

```
docker run -d -p 1025:1025 -p 8025:8025 mailhog/mailhog
```

then go to http://localhost:8025/ to see sended email (mailhog catch all, so you can send mail to anything without concern).

**Comment by gbonaspetti** [comment]
Tested on branch.

**Comment by PhilippeAccorsi** [comment]
@inkhey can you add tracim version where this issue is fixed ?
And add the issue in the correct project is still exist.

**Comment by inkhey** [comment]
> @inkhey can you add tracim version where this issue is fixed ?
> And add the issue in the correct project is still exist.

This issue was fixed in tracim 3.1.0.
